### PR TITLE
feat: add tokenfactory module

### PR DESCRIPTION
**Pull request by hugjobk**
# Description

Closes: #85

---

## Author Checklist

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.

I have...

- [x] tackled an existing issue or discussed with a team member
- [x] left instructions on how to review the changes
- [x] targeted the correct branch (see [PR Targeting](https://github.com/titantkx/titan/blob/develop/CONTRIBUTING.md#pr-targeting))

## Reviewers Checklist

**All** items are required.
Please add a note if the item is not applicable
and please add your handle next to the items reviewed
if you only reviewed selected items.

I have...

- [x] confirmed all author checklist items have been addressed
- [x] reviewed content
- [ ] confirmed all CI checks have passed

### CODE REVIEW

File: app/app.go [hunk 1]
```diff
@@ -291,9 +299,10 @@ type App struct {
 	memKeys map[string]*storetypes.MemoryStoreKey
 
 	// keepers
-	AccountKeeper         authkeeper.AccountKeeper
-	AuthzKeeper           authzkeeper.Keeper
-	BankKeeper            bankkeeper.Keeper
+	AccountKeeper authkeeper.AccountKeeper
+	AuthzKeeper   authzkeeper.Keeper
+	// BankKeeper            bankkeeper.Keeper
```
Comment (InnerPeace080) [inline app/app.go:304]: Is it necessary to keep ?

Comment (hugjobk) [reply app/app.go]: moved to testutil

File: utils/config.go [hunk 2]
```diff
@@ -16,6 +16,8 @@ const (
 	TestnetChainID = "titan_18889"
 	// DisplayDenom defines the denomination displayed to users in client applications.
 	DisplayDenom = "tkx"
+
+	SecondaryDenom = "token"
```
[the remark this answers is not in the document]
Comment (hugjobk) [reply utils/config.go]: moved to testutil

File: x/tokenfactory/keeper/create_denom.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+			Name:    denom,
+			Symbol:  denom,
+			Display: denom,
+		}
+
+		k.bankKeeper.SetDenomMetaData(ctx, denomMetaData)
+	}
+
+	authorityMetadata := types.DenomAuthorityMetadata{
+		Admin: creatorAddr,
```
Comment (InnerPeace080) [inline x/tokenfactory/keeper/create_denom.go:48]: Does it safe when allow Admin of token free to set metadata of token?

What if someone create a token and then try to fake USDC token

Comment (hugjobk) [reply x/tokenfactory/keeper/create_denom.go:48]: Token metadata is only for displaying purpose. User should be able to verify token by denom name

File: CHANGELOG.md [hunk 4]
```diff
@@ -48,6 +48,9 @@ Ref: https://keepachangelog.com/en/1.1.0/
 
 - (amino) [#73](https://github.com/titantkx/titan/issues/73) Add regis amino codec of `nftmint` and `validatorreward` module for `gov`,`authz`,`group`.
 
+- (tokenfactory) [#88](https://github.com/titantkx/titan/pull/88) Add tokenfactory module.
+
```
Comment (InnerPeace080) [inline CHANGELOG.md:52]: This change should be in State Machine Breaking

File: tests/integration/tokenfactory/keeper/testcontracts/contracts/infinite-track-beforesend/.github/workflows/Release.yml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        env:
+          GITHUB_TOKEN: ${{ github.token }}
+      - name: Upload optimized wasm
+        uses: svenstaro/upload-release-action@v2
+        with:
+          repo_token: ${{ secrets.GITHUB_TOKEN }}
+          file: ./artifacts/*.wasm
+          tag: ${{ github.ref }}
+          overwrite: true
+          file_glob: true
```
Comment (InnerPeace080) [inline tests/integration/tokenfactory/keeper/testcontracts/contracts/infinite-track-beforesend/.github/workflows/Release.yml:35]: remove this workflow file if not needed

File: tests/integration/tokenfactory/keeper/testcontracts/contracts/infinite-track-beforesend/.github/workflows/Basic.yml [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+      - name: Generate Schema
+        uses: actions-rs/cargo@v1
+        with:
+          command: schema
+          args: --locked
+
+      - name: Schema Changes
+        # fails if any changes not committed
+        run: git diff --exit-code schema
```
Comment (InnerPeace080) [inline tests/integration/tokenfactory/keeper/testcontracts/contracts/infinite-track-beforesend/.github/workflows/Basic.yml:75]: remove this workflow file if not needed

File: x/tokenfactory/keeper/msg_server.go [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.TypeMsgSetBeforeSendHook,
+			sdk.NewAttribute(types.AttributeDenom, msg.GetDenom()),
+			sdk.NewAttribute(types.AttributeBeforeSendHookAddress, msg.GetCosmwasmAddress()),
+		),
+	})
+
+	return &types.MsgSetBeforeSendHookResponse{}, nil
+}
```
Comment (InnerPeace080) [inline x/tokenfactory/keeper/msg_server.go:233]: This message allow admin of token free to set Before send hook of that token.

Have a posible scenario that can prevent validator and delegator withdraw their reward (or increase gas price)

Step:
1. Deploy contract to create new native token
2. Regis beforesendhook
3. Send new token to Fee collected module account
4. In next block that token be distribute to all validator , delegator and  community account
5. When validator/delegator try to withdraw their reward (that it self include malicious token). It then can block withdraw transaction

File: x/tokenfactory/keeper/msg_server.go [hunk 8]
```diff
@@ -0,0 +1,233 @@
+package keeper
+
+import (
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	"github.com/titantkx/titan/x/tokenfactory/types"
+)
+
```
Comment (InnerPeace080) [inline x/tokenfactory/keeper/msg_server.go:10]: Missing support `UpdateParams` that allow config tokenfactory params by gov module.

REF: like distribution module in sdk : https://github.com/tokenize-titan/cosmos-sdk/blob/bbf75c8970e38bd7cfee84ad064ebbf18775fdfc/x/distribution/keeper/msg_server.go#L118

File: x/tokenfactory/keeper/msg_server.go (hunk 7 above)
Comment (hugjobk) [reply x/tokenfactory/keeper/msg_server.go]: Removed before send hooks

File: x/tokenfactory/keeper/msg_server.go (hunk 8 above)
Comment (hugjobk) [reply x/tokenfactory/keeper/msg_server.go:6]: Added

File: x/tokenfactory/types/errors.go [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+	ErrDenomExists              = errorsmod.Register(ModuleName, 2, "attempting to create a denom that already exists (has bank metadata)")
+	ErrUnauthorized             = errorsmod.Register(ModuleName, 3, "unauthorized account")
+	ErrInvalidDenom             = errorsmod.Register(ModuleName, 4, "invalid denom")
+	ErrInvalidCreator           = errorsmod.Register(ModuleName, 5, "invalid creator")
+	ErrInvalidAuthorityMetadata = errorsmod.Register(ModuleName, 6, "invalid authority metadata")
+	ErrInvalidGenesis           = errorsmod.Register(ModuleName, 7, "invalid genesis")
+	ErrSubdenomTooLong          = errorsmod.Register(ModuleName, 8, fmt.Sprintf("subdenom too long, max length is %d bytes", MaxSubdenomLength))
+	ErrCreatorTooLong           = errorsmod.Register(ModuleName, 9, fmt.Sprintf("creator too long, max length is %d bytes", MaxCreatorLength))
+	ErrDenomDoesNotExist        = errorsmod.Register(ModuleName, 10, "denom does not exist")
+	ErrBeforeSendHookOutOfGas   = errorsmod.Register(ModuleName, 12, "gas meter hit maximum limit")
```
Comment (InnerPeace080) [inline x/tokenfactory/types/errors.go:22]: Remove unused Error

**Review by InnerPeace080 — APPROVED** [review]

**Pull request merged**